core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -204,7 +197,62 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
         return topicPartitionData;
     }
 
-    private void releasePartitionLocks(String groupId, Set<TopicIdPartition> topicIdPartitions) {
+    /**
+     * Prepare partitions fetch data structure for acquirable partitions in the share fetch request satisfying minBytes criteria.
+     */
+    Map<TopicIdPartition, FetchPartitionData> replicaManagerFetchData(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData,
+                                                                      boolean hasRequestTimedOut) {
+        log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}", topicPartitionData,
+            shareFetchData.groupId(), shareFetchData.fetchParams());
+        boolean minBytesSatisfied = false;
+        Map<TopicIdPartition, FetchPartitionData> responseData = new HashMap<>();
+        try {
+            Seq<Tuple2<TopicIdPartition, LogReadResult>> responseLogResult = replicaManager.readFromLog(
```

Comment:
1. This approach is ok, but probably not the most efficient. `replicaManager.readFromLog` is relatively expensive. To avoid calling it on every HWM change, the DelayedFetch maintains the file position of the fetch offset and compares it with the file position of HWM to estimate the fetchable bytes. We could potentially do the same thing here. This requires us to maintain the file position for SharePartition.endOffset.
2. Ideally, we need to take into account the size of those non-acquirable batches in SharePartition when estimating the fetchable bytes.

Reply:
Thanks a lot for suggestion @junrao. This is good. 

I have aquestion on 2.
> Ideally, we need to take into account the size of those non-acquirable batches in SharePartition when estimating the fetchable bytes.

Though it's an ideal solution but the next fetch offset being prior to endOffset should be rare i.e. when some records are released or timedout. So I think we can avoid calculating non-acquirable and proceed to fetch anyways if our criteria from end offset to HWM meets. We can have the min bytes check after the fetch as currently in the PR. Wdyt?

Reply:
> Though it's an ideal solution but the next fetch offset being prior to endOffset should be rare i.e. when some records are released or timedout. So I think we can avoid calculating non-acquirable and proceed to fetch anyways if our criteria from end offset to HWM meets. We can have the min bytes check after the fetch as currently in the PR. Wdyt?

Yes, this sounds reasonable. We can just ignore the non-acquirable records for now.

> You mean the file position of the latest offset that was fetched for the share partition, right?

Basically, we need to maintain endOffset as a `LogOffsetMetadata`, which contains segment position. We can then use `LogOffsetMetadata.positionDiff` to calculate the available bytes.